### ISSUE TITLE
Support for JSONFormer or similar

### ISSUE THREAD

**Issue by pachacamac** [simonw/llm#64]
Thought it would be nice to be able to force JSON output out of box with something like https://github.com/1rgs/jsonformer or similar. Wdyt?

**Comment by simonw** [comment]
This is a really good idea. I'm not sure exactly what this would look like - ideally it's something that a LLM plugin could implement using a `Model` subclass as seen in the branch I'm developing at the moment:

- #65

**Comment by fblissjr** [comment]
Not sure if you've gotten to it yet, but I'd use something like Guidance (https://github.com/microsoft/guidance).

I actually have this half working locally now with a local LLM running in openai mode (ooba extension), and it's working amazingly well. Amazing app @simonw , I avoided using it before because I'm trying to use local models for anything shell-based, but this is a fantastic, simple app. Love the ability to pipe things in, and feel in control of what's happening.

**Comment by pachacamac** [comment]
Hey @fblissjr thanks for sharing. This is a really cool lib and definitely worth to be added as a plugin on its own (in my opinion at least) but I fail to understand how this would replace something like `jsonformer`. Could you maybe give an example how this could be used to produce a specific output format? Based on the first example in the readme it just transforms already known input.

**Comment by fblissjr** [comment]
I'm saying guidance can act as a superset to jsonformer (given that only solves json). The problem is that it doesn't work as efficiently extent without the ability to modify the logits processor. So it works great on local models, but requires more back and forth with something like OpenAI.

See guidance's jsonformer example: https://github.com/microsoft/guidance/blob/main/notebooks/applications/jsonformer.ipynb

And another example on following syntax: https://github.com/microsoft/guidance/blob/main/notebooks/guaranteeing_valid_syntax.ipynb

**Comment by pachacamac** [comment]
@fblissjr aha gotcha! Thanks for clarifying. Super interesting I'll definitely play around more with those examples.

**Comment by simonw** [comment]
See:
- #776
